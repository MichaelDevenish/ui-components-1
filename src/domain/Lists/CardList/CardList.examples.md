#### Card List with all the components

```jsx
import { Button } from '@Domain/Buttons';
import { GridLayout } from '@Domain/Layouts';
import { Variables } from '@Common';

initialState = { card1: false, card2: false, allCards: false };

const text = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur nec libero et libero molestie eleifend. Donec dignissim vel erat eu cursus.';
 
<div>
  <Button
    type='neutral'
    onClick={() => setState({ card1: !state.allCards, card2: !state.allCards, allCards: !state.allCards})}>
    {state.allCards ? 'Collapse All' : 'Expand All'}
  </Button>
  <CardList
  onCardToggle={(card) => setState({ [card.name]: !state[card.name]})}
  cards={[
    {
      name:'card1',
      isExpanded: state.card1,
      actions:[
          {
            text: 'homepage',
            href: 'https://intellihr.github.io/ui-components/',
            stopPropagation: true
          },
          {
            text: 'Delete',
            onClick: (event) => { alert('Delete action for the card 1') },
            sectionType: 'alert',
            stopPropagation: true
          }
        ],
      collapseComponent: (
        <GridLayout
          gutterPaddingX={Variables.Spacing.sSmall}
          gutterPaddingY={Variables.Spacing.sSmall}
          cells={[
            {
              size: 7,
              content: text
            },
            {
              size: 5,
              content: text
            }
          ]}
        />
      ),
      expandComponent: (
        <GridLayout
          gutterPaddingX={Variables.Spacing.sSmall}
          gutterPaddingY={Variables.Spacing.sSmall}
          cells={[
            {
              size: 12,
              content: text
            },
            {
              size: 12,
              content: text
            }
          ]}
        />
      )
    },
    {
      name:'card2',
      isExpanded: state.card2,
      actions:[
                {
                  text: 'google',
                  href: 'https://www.google.com.au',
                  stopPropagation: true
                },
                {
                  text: 'Delete',
                  onClick: (event) => { alert('Delete action for the card 2') },
                  sectionType: 'alert',
                  stopPropagation: true
                }
              ],
      collapseComponent: (
        <GridLayout
          gutterPaddingX={Variables.Spacing.sSmall}
          gutterPaddingY={Variables.Spacing.sSmall}
          cells={[
            {
              size: 7,
              content: text
            },
            {
              size: 5,
              content: text
            }
          ]}
        />
      ),
      expandComponent: (
        <GridLayout
          gutterPaddingX={Variables.Spacing.sSmall}
          gutterPaddingY={Variables.Spacing.sSmall}
          cells={[
            {
              size: 12,
              content: text
            },
            {
              size: 12,
              content: text
            }
          ]}
        />
      )
    }
  ]}
  />
</div>
```

#### Card List without some component

```jsx
import { Button } from '@Domain/Buttons';
import { GridLayout } from '@Domain/Layouts';
import { Variables } from '@Common';

initialState = { card: false, card2: false, card3: false, allCards: false };

const text = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur nec libero et libero molestie eleifend. Donec dignissim vel erat eu cursus.';
 
<div>
  <Button
    type='neutral'
    onClick={() => setState({ card1: !state.allCards, card2: !state.allCards, allCards: !state.allCards})}>
    {state.allCards ? 'Collapse All' : 'Expand All'}
  </Button>
  <CardList
  onCardToggle={(card) => setState({ [card.name]: !state[card.name]})}
  cards={[
    {
      name:'card1',
      isExpanded: state.card1,
      collapseComponent: <div><div>I have expand component only</div>{text}</div>,
      expandComponent: (
        <GridLayout
          gutterPaddingX={Variables.Spacing.sSmall}
          gutterPaddingY={Variables.Spacing.sSmall}
          cells={[
            {
              size: 12,
              content: text
            },
            {
              size: 12,
              content: text
            }
          ]}
        />
      )
    },
    {
      name:'card2',
      isExpanded: state.card2,
      collapseComponent: <div><div>I don't have expand component and actions</div>{text}</div>,
    },
    {
      name:'card3',
      isExpanded: state.card3,
      actions:[
                {
                  text: 'google',
                  href: 'https://www.google.com.au',
                  stopPropagation: true
                },
                {
                  text: 'Delete',
                  onClick: (event) => { alert('Delete action for the card 2') },
                  sectionType: 'alert',
                  stopPropagation: true
                }
              ],
      collapseComponent: <div><div>I have actions only</div>{text}</div>,
    }
  ]}
  />
</div>
```
#### Card List skeleton

```jsx

<CardList
    showSkeleton
    cards={[]}
    />
```
