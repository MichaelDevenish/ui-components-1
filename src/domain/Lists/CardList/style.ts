import styled, { css, keyframes } from 'styled-components'

import { Variables } from '../../../common'

export interface IExpandComponentWrapperProps {
  isExpanded: boolean
}

export interface ICardWrapperProps {
  isExpanded: boolean,
  hasHoverStyle: boolean
}

export interface IStyleToggleButtonProps {
  isExpanded: boolean,
  hasParentHoverStyle: boolean
}

const CardWrapper = styled.div`
  height: auto;
  margin-bottom: 8px;
  padding: ${Variables.Spacing.sMedium}px ${Variables.Spacing.sMedium}px ${Variables.Spacing.sMedium}px ${Variables.Spacing.sMedium}px
  background-color: ${Variables.Color.n100};
  border: 1px solid ${Variables.Color.n250};
  border-radius: ${Variables.Style.borderRadius}px;
  transition: height .5s;

  ${(props: ICardWrapperProps) => props.hasHoverStyle && css`
      cursor: pointer;
      &:hover {
      background-color: ${Variables.Color.n200};
      transition: background-color .25s ease-out;
    }
  `}
`

const ButtonStyle = css`
  outline: none;
  color: ${Variables.Color.n600};
  width: 28px;
  height: 28px;
  border-radius: 50%;

  &:hover {
    color: ${Variables.Color.n800};
    background-color: ${Variables.Color.n300};
    transition: all .25s ease-out;
  }
`

const StyleToggleButton = styled.button`
  ${ButtonStyle};

  ${(props: IStyleToggleButtonProps) => props.isExpanded && css`
      transform: rotate(180deg);
  `}

  ${(props: IStyleToggleButtonProps) => props.hasParentHoverStyle && css`
    ${CardWrapper}:hover & {
      color: ${Variables.Color.n800};
      background-color: ${Variables.Color.n300};
    }
  `}
`

const StyleActionButton = styled.button`
  ${ButtonStyle};

  &:hover ${StyleToggleButton} {
    background-color: transparent;
    color: ${Variables.Color.n600};
  }

  &:hover ${CardWrapper} {
    background-color: ${Variables.Color.n100};
  }
`

const AnimateIn = keyframes`
  0% {
    opacity: 0;
    transform: translateY(-${Variables.Spacing.s3XSmall});
  }

  50% {
    transform: translateY(0);
  }

  100% {
    opacity: 1;
  }
`

const AnimateOut = keyframes`
  0% {
    opacity: 1;
    transform: translateY(0);
  }

  50% {
    transform: translateY(-${Variables.Spacing.s3XSmall});
  }

  100% {
    opacity: 0;
  }
`

const ExpandComponentWrapper = styled.div`
  display: none;
  animation-name: ${AnimateOut};
  animation-duration: 0.5s;
  animation-iteration-count: 1;
  animation-direction: alternate;
  animation-timing-function: ease-in-out;
  animation-fill-mode: both;
  animation-delay: 0s;

  ${(props: IExpandComponentWrapperProps) => props.isExpanded && css`
      display: block;
      animation-name: ${AnimateIn};
      animation-duration: 0.6s;
      animation-iteration-count: 1;
      animation-direction: normal;
      animation-timing-function: ease-in;
      animation-fill-mode: both;
      animation-delay: 0s;
  `}
`

export {
  CardWrapper,
  ExpandComponentWrapper,
  StyleActionButton,
  StyleToggleButton
}
