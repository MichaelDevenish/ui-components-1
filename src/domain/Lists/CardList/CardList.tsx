import React from 'react'

import { Variables } from '../../../common'
import { FontAwesomeIcon } from '../../Icons/FontAwesomeIcon'
import { GridLayout } from '../../Layouts/GridLayout'
import { DropdownMenu, IDropdownMenuToggleComponentProps } from '../../Popovers/DropdownMenu'
import { ISectionProps } from '../../Popovers/DropdownMenu/subcomponents/Section'
import { BlockSkeleton } from '../../Skeletons/BlockSkeleton'
import { CardWrapper, ExpandComponentWrapper, StyleActionButton, StyleToggleButton } from './style'

interface ICardProps {
  name: string
  /** the component that shows even in collapse mode */
  collapseComponent: JSX.Element
  /** the component that shows in expand mode only */
  expandComponent?: JSX.Element
  /** If true, the card is in expand style */
  isExpanded?: boolean
  /** Array of actions to display in all the cards action button popover */
  actions?: ISectionProps[]
}

interface ICardListProps {
  /** Array of cards to display in the list */
  cards: ICardProps[]
  /** callback when user toggle card */
  onCardToggle?: (card: ICardProps) => void
  /** If true, the card list shows its skeleton */
  showSkeleton?: boolean
}

interface ICardListState {
  isActionButtonHover: boolean
}

export class CardList extends React.PureComponent<ICardListProps,ICardListState> {
  public static defaultProps: Partial<ICardListProps> = {
    showSkeleton: false
  }

  public state: ICardListState = {
    isActionButtonHover: false
  }

  public render (): JSX.Element {
    const {
      cards,
      showSkeleton
    } = this.props

    if (showSkeleton) {
      return (
        <>
          <BlockSkeleton
            showSkeleton
            height={100}
          />
          <BlockSkeleton
            showSkeleton
            height={100}
          />
          <BlockSkeleton
            showSkeleton
            height={100}
          />
        </>
      )
    }

    return (
      <>
        {cards.map((card, index) => this.cardContent(card, index))}
      </>
    )
  }

  private cardContent = (card: ICardProps, index: number) => {
    const {
      collapseComponent,
      expandComponent,
      isExpanded,
      name
    } = card

    return(
      <CardWrapper
        onClick={this.onToggleCard(card)}
        key={`${name}-${index}`}
        isExpanded={!!isExpanded}
        hasHoverStyle={!this.state.isActionButtonHover && !!card.expandComponent}
      >
        <GridLayout
          gutterPaddingX={Variables.Spacing.sSmall}
          cells={[
            {
              size: 'auto',
              content: collapseComponent
            },
            {
              size: 'shrink',
              content: this.cardRightComponent(card),
              displayType: 'flex'
            }
          ]}
        />
        <ExpandComponentWrapper isExpanded={!!isExpanded}>
          {expandComponent}
        </ExpandComponentWrapper>
      </CardWrapper>
    )
  }

  private onToggleCard = (card: ICardProps) => () => {
    const {
      onCardToggle
    } = this.props

    if (onCardToggle) {
      onCardToggle(card)
    }
  }

  private cardRightComponent = (card: ICardProps) => {
    const gridToggleButtonCell = {
      size: 'shrink' as 'shrink',
      content: this.toggleButton(card),
      displayType: 'flex' as 'flex'
    }

    const gridActionButtonDropdownMenuCell = {
      size: 'shrink'as 'shrink',
      content: this.actionButtonDropdownMenu(card),
      displayType: 'flex' as 'flex'
    }

    const cardRightComponentCells = []

    if (this.actionButtonDropdownMenu(card)) {
      cardRightComponentCells.push(gridActionButtonDropdownMenuCell)
    }

    if (this.toggleButton(card)) {
      cardRightComponentCells.push(gridToggleButtonCell)
    }

    if (cardRightComponentCells.length > 0) {
      return(
        <GridLayout
          horizontalAlignment={GridLayout.HorizontalAlignment.Justify}
          gutterPaddingX={Variables.Spacing.sSmall}
          cells={cardRightComponentCells}
        />
      )
    }

    return null
  }

  private actionButtonDropdownMenu = (card: ICardProps) => {
    const {
      actions
    } = card

    if (actions) {
      return(
        <DropdownMenu
          sections={actions}
          toggleComponent={this.actionButton}
        />
      )
    }

    return null
  }

  private actionButton = ({ toggleMenu, toggleComponentRef}: IDropdownMenuToggleComponentProps) => (
    <div>
      <StyleActionButton
        onClick={this.handleActionButtonClick(toggleMenu)}
        onMouseOver={this.handleActionButtonMouseOver}
        onMouseOut={this.handleActionButtonMouseOut}
        innerRef={toggleComponentRef}
      >
        <FontAwesomeIcon type='ellipsis-v'/>
      </StyleActionButton>
    </div>
  )

  private toggleButton = (card: ICardProps) => {
    if (card.expandComponent) {
      return(
        <div>
          <StyleToggleButton isExpanded={!!card.isExpanded} hasParentHoverStyle={!this.state.isActionButtonHover}>
            <FontAwesomeIcon type='chevron-down'/>
          </StyleToggleButton>
        </div>
      )
    }
  }

  private handleActionButtonClick = (toggleMenu: () => void) => (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation()
    toggleMenu()
  }

  private handleActionButtonMouseOver = () => {
    this.setState({ isActionButtonHover: true })
  }

  private handleActionButtonMouseOut = () => {
    this.setState({ isActionButtonHover: false })
  }
}
