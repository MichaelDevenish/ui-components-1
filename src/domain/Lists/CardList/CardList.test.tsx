import { shallow } from 'enzyme'
import React from 'react'

import { CardList } from './CardList'

const dummyFunction = () => (alert('dummy'))

describe('<CardList />', () => {
  it('should render a simple card list with expand component and actions', () => {
    const wrapper = shallow(
      <CardList
        onCardToggle={dummyFunction}
        cards={[
          {
            name: 'test1',
            collapseComponent: <div>test1</div>,
            expandComponent: <div>test1</div>,
            actions: [
              {
                text: 'action1',
                href: 'https://www.google.com.au',
              },
              {
                text: 'action2',
                href: 'https://www.google.com.au',
              }
            ]
          }
        ]}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render a simple card list with expand component only', () => {
    const wrapper = shallow(
      <CardList
        onCardToggle={dummyFunction}
        cards={[
          {
            name: 'test2',
            collapseComponent: <div>test2</div>,
            expandComponent: <div>test2</div>,
          }
        ]}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render a simple card list with actions only', () => {
    const wrapper = shallow(
      <CardList
        onCardToggle={dummyFunction}
        cards={[
          {
            name: 'test3',
            collapseComponent: <div>test3</div>,
            actions: [
              {
                text: 'action1',
                href: 'https://www.google.com.au',
              },
              {
                text: 'action2',
                href: 'https://www.google.com.au',
              }
            ]
          }
        ]}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render a simple card list without expand component and actions', () => {
    const wrapper = shallow(
      <CardList
        onCardToggle={dummyFunction}
        cards={[
          {
            name: 'test4',
            collapseComponent: <div>test4</div>,
          }
        ]}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render a card list skeleton', () => {
    const wrapper = shallow(
      <CardList
        showSkeleton
        cards={[]}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })
})
