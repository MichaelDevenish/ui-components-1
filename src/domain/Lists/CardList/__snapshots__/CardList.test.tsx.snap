// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardList /> should render a card list skeleton 1`] = `
<Fragment>
  <BlockSkeleton
    height={100}
    showSkeleton={true}
  />
  <BlockSkeleton
    height={100}
    showSkeleton={true}
  />
  <BlockSkeleton
    height={100}
    showSkeleton={true}
  />
</Fragment>
`;

exports[`<CardList /> should render a simple card list with actions only 1`] = `
<Fragment>
  <styled.div
    hasHoverStyle={false}
    isExpanded={false}
    key="test3-0"
    onClick={[Function]}
  >
    <GridLayout
      cells={
        Array [
          Object {
            "content": <div>
              test3
            </div>,
            "size": "auto",
          },
          Object {
            "content": <GridLayout
              cells={
                Array [
                  Object {
                    "content": <DropdownMenu
                      hasInitialFocus={false}
                      sections={
                        Array [
                          Object {
                            "href": "https://www.google.com.au",
                            "text": "action1",
                          },
                          Object {
                            "href": "https://www.google.com.au",
                            "text": "action2",
                          },
                        ]
                      }
                      toggleComponent={[Function]}
                    />,
                    "displayType": "flex",
                    "size": "shrink",
                  },
                ]
              }
              gridColumns={12}
              gutterMarginX="none"
              gutterMarginY="none"
              gutterPaddingX={12}
              gutterPaddingY="none"
              horizontalAlignment="justify"
              verticalAlignment="stretch"
            />,
            "displayType": "flex",
            "size": "shrink",
          },
        ]
      }
      gridColumns={12}
      gutterMarginX="none"
      gutterMarginY="none"
      gutterPaddingX={12}
      gutterPaddingY="none"
      horizontalAlignment="left"
      verticalAlignment="stretch"
    />
    <styled.div
      isExpanded={false}
    />
  </styled.div>
</Fragment>
`;

exports[`<CardList /> should render a simple card list with expand component and actions 1`] = `
<Fragment>
  <styled.div
    hasHoverStyle={true}
    isExpanded={false}
    key="test1-0"
    onClick={[Function]}
  >
    <GridLayout
      cells={
        Array [
          Object {
            "content": <div>
              test1
            </div>,
            "size": "auto",
          },
          Object {
            "content": <GridLayout
              cells={
                Array [
                  Object {
                    "content": <DropdownMenu
                      hasInitialFocus={false}
                      sections={
                        Array [
                          Object {
                            "href": "https://www.google.com.au",
                            "text": "action1",
                          },
                          Object {
                            "href": "https://www.google.com.au",
                            "text": "action2",
                          },
                        ]
                      }
                      toggleComponent={[Function]}
                    />,
                    "displayType": "flex",
                    "size": "shrink",
                  },
                  Object {
                    "content": <div>
                      <styled.button
                        hasParentHoverStyle={true}
                        isExpanded={false}
                      >
                        <FontAwesomeIcon
                          type="chevron-down"
                        />
                      </styled.button>
                    </div>,
                    "displayType": "flex",
                    "size": "shrink",
                  },
                ]
              }
              gridColumns={12}
              gutterMarginX="none"
              gutterMarginY="none"
              gutterPaddingX={12}
              gutterPaddingY="none"
              horizontalAlignment="justify"
              verticalAlignment="stretch"
            />,
            "displayType": "flex",
            "size": "shrink",
          },
        ]
      }
      gridColumns={12}
      gutterMarginX="none"
      gutterMarginY="none"
      gutterPaddingX={12}
      gutterPaddingY="none"
      horizontalAlignment="left"
      verticalAlignment="stretch"
    />
    <styled.div
      isExpanded={false}
    >
      <div>
        test1
      </div>
    </styled.div>
  </styled.div>
</Fragment>
`;

exports[`<CardList /> should render a simple card list with expand component only 1`] = `
<Fragment>
  <styled.div
    hasHoverStyle={true}
    isExpanded={false}
    key="test2-0"
    onClick={[Function]}
  >
    <GridLayout
      cells={
        Array [
          Object {
            "content": <div>
              test2
            </div>,
            "size": "auto",
          },
          Object {
            "content": <GridLayout
              cells={
                Array [
                  Object {
                    "content": <div>
                      <styled.button
                        hasParentHoverStyle={true}
                        isExpanded={false}
                      >
                        <FontAwesomeIcon
                          type="chevron-down"
                        />
                      </styled.button>
                    </div>,
                    "displayType": "flex",
                    "size": "shrink",
                  },
                ]
              }
              gridColumns={12}
              gutterMarginX="none"
              gutterMarginY="none"
              gutterPaddingX={12}
              gutterPaddingY="none"
              horizontalAlignment="justify"
              verticalAlignment="stretch"
            />,
            "displayType": "flex",
            "size": "shrink",
          },
        ]
      }
      gridColumns={12}
      gutterMarginX="none"
      gutterMarginY="none"
      gutterPaddingX={12}
      gutterPaddingY="none"
      horizontalAlignment="left"
      verticalAlignment="stretch"
    />
    <styled.div
      isExpanded={false}
    >
      <div>
        test2
      </div>
    </styled.div>
  </styled.div>
</Fragment>
`;

exports[`<CardList /> should render a simple card list without expand component and actions 1`] = `
<Fragment>
  <styled.div
    hasHoverStyle={false}
    isExpanded={false}
    key="test4-0"
    onClick={[Function]}
  >
    <GridLayout
      cells={
        Array [
          Object {
            "content": <div>
              test4
            </div>,
            "size": "auto",
          },
          Object {
            "content": null,
            "displayType": "flex",
            "size": "shrink",
          },
        ]
      }
      gridColumns={12}
      gutterMarginX="none"
      gutterMarginY="none"
      gutterPaddingX={12}
      gutterPaddingY="none"
      horizontalAlignment="left"
      verticalAlignment="stretch"
    />
    <styled.div
      isExpanded={false}
    />
  </styled.div>
</Fragment>
`;
